flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/LogicalWindowAggregateRule.scala

```diff
@@ -117,46 +119,86 @@ class LogicalWindowAggregateRule
   }
 
   private def identifyWindow(field: RexNode): Option[Window] = {
-    // Detects window expressions by pattern matching
-    //   supported patterns: FLOOR(time AS xxx) and CEIL(time AS xxx),
-    //   with time being equal to proctime() or rowtime()
     field match {
       case call: RexCall =>
         call.getOperator match {
-          case _: SqlFloorFunction =>
-            val operand = call.getOperands.get(1).asInstanceOf[RexLiteral]
-            val unit: TimeUnitRange = operand.getValue.asInstanceOf[TimeUnitRange]
-            val w = LogicalWindowAggregateRule.timeUnitRangeToTumbleWindow(unit)
-            call.getType match {
-              case TimeModeTypes.PROCTIME =>
-                return Some(w)
-              case TimeModeTypes.ROWTIME =>
-                return Some(w.on("rowtime"))
-              case _ =>
-            }
-          case _ =>
+          case _: SqlFloorFunction => FloorWindowTranslator(call).toWindow
+          case SqlStdOperatorTable.TUMBLE => TumbleWindowTranslator(call).toWindow
+          case SqlStdOperatorTable.HOP => SlidingWindowTranslator(call).toWindow
+          case SqlStdOperatorTable.SESSION => SessionWindowTranslator(call).toWindow
+          case _ => None
         }
-      case _ =>
+      case _ => None
     }
-    None
   }
-
 }
 
-object LogicalWindowAggregateRule {
+private abstract class WindowTranslator {
+  val call: RexCall
 
-  private[flink] val LOGICAL_WINDOW_PREDICATE = RelOptRule.operand(classOf[LogicalAggregate],
-    RelOptRule.operand(classOf[LogicalProject], RelOptRule.none()))
+  protected def unwrapLiteral[T](node: RexNode): T =
+    node.asInstanceOf[RexLiteral].getValue.asInstanceOf[T]
 
-  private[flink] val INSTANCE = new LogicalWindowAggregateRule
+  protected def getOperandAsLong(idx: Int): Long =
+    unwrapLiteral[BigDecimal](call.getOperands.get(idx)).longValue()
```

Comment:
I just tried out Calcite did not stop you from passing something like a {{RexCall}}. So yes, it can be dynamic.

One question: whether Flink actually supports `GroupWindow` that has a dynamic size? Maybe I'm wrong but it does not seem so.

If the answer is no maybe we should check that whether it is a `RexLiteral`?